runtime/vm/classsupport.c

```diff
@@ -226,6 +226,24 @@ internalCreateArrayClass(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Clas
 		LOAD_LOCATION_UNKNOWN,
 		NULL,
 		NULL);
+
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+	/* When creating an array of valuetype elements, the array elements are initialized to the defaultValue of the
+	 * element type. As a result the element type must be fully initialized (if its a valuetype) before creating an
+	 * instance of the array.
+	 */
+	if ((NULL != result) && J9_IS_J9CLASS_VALUETYPE(elementClass)) {
+		UDATA initStatus = elementClass->initializeStatus;
+		if ((J9ClassInitSucceeded != initStatus) && ((UDATA)vmThread != initStatus)) {
+			initializeClass(vmThread, elementClass);
+			if (NULL != vmThread->currentException) {
+				result = NULL;
+				elementClass->arrayClass = NULL;
```

Comment:
This is problematic - there's a timing hole where the array class would be visible to another thread before being NULLed.

Probably the answer is to initialize the VT class before starting to create the array class.